### PR TITLE
Pr ekf2 param instances

### PR DESCRIPTION
**Pull request by tumbili**

### CODE REVIEW

**Comment by LorenzMeier** [comment]
@tumbili I'm not sure what the rationale for the change is, but I presume there are good reasons?

**Comment by tumbili** [comment]
@LorenzMeier I wanted to free the memory used by the parameters if ekf2 app was closed.
Furthermore, I wanted to abort ekf2 if any parameter instance couldn't be allocated.

**Comment by tumbili** [comment]
@LorenzMeier I think the code style commit blows up the diff. The real work should be in the first commit. Seems a bit ridiculous to delete all those pointers one by one, can you think of a better idea? Maybe some kind of array which we could loop through?

**Comment by LorenzMeier** [comment]
Why are those parameters not just class members instead? What is the value of using a pointer in the first place?

**Comment by tumbili** [comment]
@LorenzMeier I need to give those classes a pointer to the ekf parameter struct. Since the ekf instance does not exist before I needed to work with pointers.

**Comment by bkueng** [comment]
@tumbili You still need to check for nullptr in the descructor. But you could also do something like:
```

class Ekf2
{
	struct Ekf2Params
	{
		control::BlockParamFloat _mag_delay_ms;
		...
		//constructor:
		Ekf2Params(Ekf2& ekf, parameters& p) :
		_mag_delay_ms(ekf, "xy", false, &p.mag_delay_ms),
		... {}
	};

	Ekf2Params *_ekf2_params = nullptr;

	bool init_params() {
		_params = new ...
		if (!_params) return false
		_ekf2_params = new ...
		if (!_ekf2_params) return false
	}

```
This way you only need to delete _params & _ekf2_params and still have everything on the heap.

**Comment by LorenzMeier** [comment]
The EKF instance does exist in the constructor after the line where you call it. No need to do what you're trying to do, just use the normal class members and initialisers.

File: src/modules/ekf2/ekf2_main.cpp [hunk 1]
```diff
@@ -263,79 +283,158 @@ Ekf2::Ekf2():
 	_lp_roll_rate(250.0f, 30.0f),
 	_lp_pitch_rate(250.0f, 30.0f),
 	_lp_yaw_rate(250.0f, 20.0f),
-	_ekf(new Ekf()),
```
Comment (LorenzMeier) [inline src/modules/ekf2/ekf2_main.cpp]: @tumbili What is the problem with the structure here? This is completely valid C++

**Comment by tumbili** [comment]
@LorenzMeier Wasn't sure if it's guaranteed that the instance exists.
@bkueng Thanks for the help!

**Comment by bkueng** [comment]
Yes _ekf is initialized before. The problem is that _ekf can ben nullptr, and it is dereferenced for _params. My solution handles this. Making _ekf a non-pointer would solve this too.

**Comment by LorenzMeier** [comment]
Just turn EKF2 into a class member. There is no reason for it to be a pointer.

**Comment by LorenzMeier** [comment]
We really should not perform memory management manually.

**Comment by tumbili** [comment]
@LorenzMeier I removed all pointers and made all instances class members. Tested in SITL.

**Comment by LorenzMeier** [comment]
Nice! Now it really reads like sensible C++. You now just need to check if the initial instantiation fails of the EKF2 wrapper (maybe this already is in place). If the system is out of memory it will fail right there.

**Comment by tumbili** [comment]
Merged!

**Pull request closed without merging**